File: examples/02-kprobe-counter/python/counter.py
#!/usr/bin/env python3
"""Equivalente em BCC de counter.bpf.c/counter.c (../src): soma o argumento
"size" de cada chamada de tcp_sendmsg() em um contador global. Em vez de
receber os argumentos já tipados (o que exigiria puxar os cabeçalhos
completos do kernel para definir "struct sock"/"struct msghdr" -- desnecessário
aqui, já que só usamos "size"), lemos o 3º argumento diretamente do registrador
via PT_REGS_PARM3, seguindo a convenção de chamada x86-64 (sk, msg, size ->
rdi, rsi, rdx). É o estilo "clássico" do BCC, mais leve que o equivalente
tipado usado no exemplo 04 (que precisa mesmo de "struct sock" completo).

Uso: sudo python3 counter.py
"""
from time import sleep

from bcc import BPF

programa = r"""
#include <asm/ptrace.h>

BPF_ARRAY(bytes_totais, u64, 1);

int ao_enviar(struct pt_regs *ctx) {
    size_t size = (size_t)PT_REGS_PARM3(ctx);
    int chave = 0;
    u64 *total = bytes_totais.lookup(&chave);
    u64 novo_total = total ? *total + size : size;
    bytes_totais.update(&chave, &novo_total);
    return 0;
}
"""

if __name__ == "__main__":
    b = BPF(text=programa)
    b.attach_kprobe(event="tcp_sendmsg", fn_name="ao_enviar")

    print("Contando bytes enviados via TCP em todo o sistema... Ctrl+C para sair.")
    try:
        while True:
            total = b["bytes_totais"][0].value
            print(f"\rbytes enviados (tcp_sendmsg) desde o inicio: {total}", end="", flush=True)
            sleep(1)
    except KeyboardInterrupt:
        print()

File: examples/02-kprobe-counter/src/counter.c
// Idêntico em estrutura ao loader do exemplo 01 — o ciclo libbpf (abrir,
// carregar, anexar, ler o mapa) é o mesmo independentemente do tipo de hook.
#include <signal.h>
#include <stdio.h>
#include <unistd.h>

#include <bpf/bpf.h>
#include <bpf/libbpf.h>

static volatile sig_atomic_t parar;

static void parar_ao_receber_sinal(int sig)
{
    (void)sig;
    parar = 1;
}

int main(void)
{
    struct bpf_object *obj;
    struct bpf_program *prog;
    struct bpf_link *link;
    int map_fd;
    __u32 chave = 0;
    __u64 total;

    signal(SIGINT, parar_ao_receber_sinal);

    obj = bpf_object__open_file("build/counter.bpf.o", NULL);
    if (!obj) {
        fprintf(stderr, "erro ao abrir build/counter.bpf.o\n");
        return 1;
    }

    if (bpf_object__load(obj)) {
        fprintf(stderr, "erro ao carregar/verificar o programa (rode como root?)\n");
        return 1;
    }

    prog = bpf_object__find_program_by_name(obj, "ao_enviar");
    if (!prog) {
        fprintf(stderr, "programa 'ao_enviar' nao encontrado no objeto\n");
        return 1;
    }

    link = bpf_program__attach(prog);
    if (!link) {
        fprintf(stderr, "erro ao anexar a kprobe/tcp_sendmsg\n");
        return 1;
    }

    map_fd = bpf_object__find_map_fd_by_name(obj, "bytes_totais");
    if (map_fd < 0) {
        fprintf(stderr, "mapa 'bytes_totais' nao encontrado\n");
        return 1;
    }

    printf("Contando bytes enviados via TCP em todo o sistema... Ctrl+C para sair.\n");
    while (!parar) {
        if (bpf_map_lookup_elem(map_fd, &chave, &total) == 0)
            printf("\rbytes enviados (tcp_sendmsg) desde o inicio: %llu", (unsigned long long)total);
        fflush(stdout);
        sleep(1);
    }
    printf("\n");

    bpf_link__destroy(link);
    bpf_object__close(obj);
    return 0;
}
